### PR TITLE
Improve series matching

### PR DESCRIPTION
**Pull request by fredriklindberg**
Improves series name matching (thus better hit ratio at tmdb) for paths that contain more information that just the series name (ex. show.s01). The output from the SeriesResolver is a "clean" series name.

### CODE REVIEW

File: Emby.Server.Implementations/Library/LibraryManager.cs [hunk 1]
```diff
@@ -2688,6 +2688,21 @@ public NamingOptions GetNamingOptions()
...
+            var episodeParser = new EpisodeResolver(new NamingOptions());
+            var episodeInfo = episodeParser.Resolve(path, true, true, false, false, true);
+            if (episodeInfo == null || string.IsNullOrEmpty(episodeInfo.SeriesName))
+            {
+                return null;
+            }
+
+            return new ItemLookupInfo
+            {
+                Name = episodeInfo.SeriesName.Replace(".", " ", true, CultureInfo.InvariantCulture)
```
Comment (crobibero) [inline Emby.Server.Implementations/Library/LibraryManager.cs:2702]: What happens with a series such as [S.W.A.T.](https://www.tvmaze.com/shows/21532/swat)?

File: MediaBrowser.Providers/Plugins/Tmdb/TV/TmdbSeriesProvider.cs [hunk 2]
```diff
@@ -186,9 +186,19 @@ public async Task<MetadataResult<Series>> GetMetadata(SeriesInfo info, Cancellat
...
-                // Caller provides the filename with extension stripped and NOT the parsed filename
-                var parsedName = _libraryManager.ParseName(info.Name);
+                ItemLookupInfo parsedName;
+                // Try getting series name from the path
+                var seriesName = _libraryManager.GetSeriesFromPath(info.Path);
+                if (seriesName != null)
+                {
+                    parsedName = seriesName;
+                } else
+                {
```
Comment (crobibero) [inline MediaBrowser.Providers/Plugins/Tmdb/TV/TmdbSeriesProvider.cs:196]:
```suggestion
                }
                else
                {
```

File: Emby.Naming/Common/NamingOptions.cs [hunk 3]
```diff
@@ -137,7 +137,7 @@ public NamingOptions()
...
             {
-                @"[ _\,\.\(\)\[\]\-](3d|sbs|tab|hsbs|htab|mvc|HDR|HDC|UHD|UltraHD|4k|ac3|dts|custom|dc|divx|divx5|dsr|dsrip|dutch|dvd|dvdrip|dvdscr|dvdscreener|screener|dvdivx|cam|fragment|fs|hdtv|hdrip|hdtvrip|internal|limited|multisubs|ntsc|ogg|ogm|pal|pdtv|proper|repack|rerip|retail|cd[1-9]|r3|r5|bd5|bd|se|svcd|swedish|german|read.nfo|nfofix|unrated|ws|telesync|ts|telecine|tc|brrip|bdrip|480p|480i|576p|576i|720p|720i|1080p|1080i|2160p|hrhd|hrhdtv|hddvd|bluray|blu-ray|x264|x265|h264|xvid|xvidvd|xxx|www.www|AAC|DTS|\[.*\])([ _\,\.\(\)\[\]\-]|$)",
+                @"[ _\,\.\(\)\[\]\-](([sS][0-9]{1,3}([eE][0-9]+){0,1})|([0-9]{1,3}[xX]([0-9]+){0,1})|3d|sbs|tab|hsbs|htab|mvc|HDR|HDC|UHD|UltraHD|4k|ac3|dts|custom|dc|divx|divx5|dsr|dsrip|dutch|dvd|dvdrip|dvdscr|dvdscreener|screener|dvdivx|cam|fragment|fs|hdtv|hdrip|hdtvrip|internal|limited|multisubs|ntsc|ogg|ogm|pal|pdtv|proper|repack|rerip|retail|cd[1-9]|r3|r5|bd5|bd|se|svcd|swedish|german|read.nfo|nfofix|unrated|ws|telesync|ts|telecine|tc|brrip|bdrip|480p|480i|576p|576i|720p|720i|1080p|1080i|2160p|hrhd|hrhdtv|hddvd|bluray|blu-ray|x264|x265|h264|xvid|xvidvd|xxx|www.www|AAC|DTS|\[.*\])([ _\,\.\(\)\[\]\-]|$)",
```
Comment (crobibero) [inline Emby.Naming/Common/NamingOptions.cs:140]: Since you've edited the Regex, can you please add tests to support the new naming format?

**Comment by fredriklindberg** [comment]
I've addressed the PR comments, also re-worked the cleanstring regex further and added more tests.

This change significantly improves match rate for a new database with entries without a prior tmdb id, especially after the deprecation of tvdb.

File: Emby.Naming/Common/NamingOptions.cs [hunk 4]
```diff
@@ -137,7 +137,7 @@ public NamingOptions()
...
+                @"([ \.\,_\-]*([sS][0-9]+([eE][0-9]+)*|[0-9]+[xX][0-9]+)([ _\,\.\(\)\[\]\-][^ _\,\.\(\)\[\]\-]+)+)|([ \.\,_\-]+(\[.*\]|3d|sbs|tab|hsbs|htab|mvc|HDR|HDC|UHD|UltraHD|4k|ac3|dts|custom|dc|divx|divx5|dsr|dsrip|dutch|dvd|dvdrip|dvdscr|dvdscreener|screener|dvdivx|cam|fragment|fs|hdtv|hdrip|hdtvrip|internal|limited|multisubs|ntsc|ogg|ogm|pal|pdtv|proper|repack|rerip|retail|cd[1-9]|r3|r5|bd5|bd|se|svcd|swedish|german|read.nfo|nfofix|unrated|ws|telesync|ts|telecine|tc|brrip|bdrip|480p|480i|576p|576i|720p|720i|1080p|1080i|2160p|hrhd|hrhdtv|hddvd|bluray|blu-ray|x264|x265|h264|h265|xvid|xvidvd|xxx|www.www|hdtv|AAC|DTS|web|web-dl|amzn|hulu|ddp[0-9]|multi))+[ \.\,_\-]*",
```
Comment (cvium) [inline Emby.Naming/Common/NamingOptions.cs:140]: I don't like having series related stuff in this regex tbh. It is supposed to be a generic string cleaning regex.

File: MediaBrowser.Controller/Library/ILibraryManager.cs [hunk 5]
```diff
@@ -410,6 +410,13 @@ public interface ILibraryManager
         /// <returns><c>true</c> if [is audio file] [the specified path]; otherwise, <c>false</c>.</returns>
         bool IsAudioFile(string path);
 
+        /// <summary>
+        /// Gets the series name from path.
+        /// </summary>
+        /// <param name="path">The path.</param>
+        /// <returns>ItemInfo.</returns>
+        ItemLookupInfo GetSeriesFromPath(string path);
```
Comment (cvium) [inline MediaBrowser.Controller/Library/ILibraryManager.cs:418]: I know that we already have `GetSeasonNumberFromPath` in this interface, but I do wonder if it would not be better suited for a new `SeriesResolver`. It's pretty terrible design to have media type specific methods in the library manager imo.

Or maybe using EpisodePathParser directly in tmdb is better?

Or better yet, extending `EpisodeInfo` with the SeriesName so it's available to all providers automatically? `EpisodeInfo` is populated here https://github.com/jellyfin/jellyfin/blob/master/MediaBrowser.Controller/Entities/TV/Episode.cs#L305

File: Emby.Server.Implementations/Library/LibraryManager.cs [hunk 6]
```diff
@@ -2699,6 +2699,21 @@ public NamingOptions GetNamingOptions()
             return _namingOptions;
         }
 
+        public ItemLookupInfo GetSeriesFromPath(string path)
+        {
+            var episodeParser = new EpisodeResolver(new NamingOptions());
```
Comment (cvium) [inline Emby.Server.Implementations/Library/LibraryManager.cs:2704]: Don't create a new namingoptions. Use `GetNamingOptions()`.

File: Emby.Server.Implementations/Library/LibraryManager.cs [hunk 7]
```diff
@@ -2699,6 +2699,21 @@ public NamingOptions GetNamingOptions()
...
+            var episodeInfo = episodeParser.Resolve(path, true, true, false, false, true);
+            if (episodeInfo == null || string.IsNullOrEmpty(episodeInfo.SeriesName))
```
Comment (cvium) [inline Emby.Server.Implementations/Library/LibraryManager.cs:2706]: Could be shortened to `string.IsNullOrEmpty(episodeInfo?.SeriesName)`

**Review by cvium — CHANGES_REQUESTED** [review]

File: Emby.Naming/Common/NamingOptions.cs (hunk 4 above)
Comment (fredriklindberg) [reply Emby.Naming/Common/NamingOptions.cs:140]: Fair enough

File: MediaBrowser.Controller/Library/ILibraryManager.cs (hunk 5 above)
Comment (fredriklindberg) [reply MediaBrowser.Controller/Library/ILibraryManager.cs:418]: Yeah, I don't really like that the name have to be "cleaned" inside the provider. It should be fed a clean name, but it's not happening for the series provider right now. (it gets things like "foobar s01e1" which makes tmdb fail).

I'll see if I can figure out something better than the clean regex.

Comment (cvium) [reply MediaBrowser.Controller/Library/ILibraryManager.cs:418]: At a glance I don't see any problems with extending `GetLookupInfo` for Series, Season and Episode to include the clean series name. And shouldn't it fix the issue as well (all metadata providers will have to be updated to use it of course)?

File: Emby.Naming/TV/SeriesResolver.cs [hunk 8]
```diff
@@ -0,0 +1,64 @@
+using System.IO;
+using System.Text.RegularExpressions;
+using Emby.Naming.Common;
+
+namespace Emby.Naming.TV
+{
+    /// <summary>
+    /// Used to resolve information about series from path.
+    /// </summary>
+    public class SeriesResolver
```
Comment (cvium) [inline Emby.Naming/TV/SeriesResolver.cs:10]: This class is better suited as a static class. There's no state and only a single public method.

File: Emby.Naming/TV/SeriesPathParser.cs [hunk 9]
```diff
@@ -0,0 +1,95 @@
+using System.Globalization;
+using Emby.Naming.Common;
+
+namespace Emby.Naming.TV
+{
+    /// <summary>
+    /// Used to parse information about series from paths containing more information that only the series name.
+    /// Uses the same regular expressions as the EpisodePathParser but have different success criteria.
+    /// </summary>
+    public class SeriesPathParser
```
Comment (cvium) [inline Emby.Naming/TV/SeriesPathParser.cs:10]: I also think this should be static.

File: Emby.Naming/TV/SeriesInfo.cs [hunk 10]
```diff
@@ -0,0 +1,34 @@
...
+        /// <summary>
+        /// Gets or sets the name of the series.
+        /// </summary>
+        /// <value>The name of the series.</value>
+        public string? Name { get; set; }
+
+        /// <summary>
+        /// Gets or sets optional season number.
+        /// </summary>
+        public int? SeasonNumber { get; set; }
```
Comment (cvium) [inline Emby.Naming/TV/SeriesInfo.cs:32]: Why is this needed? Feels a bit out of place.

File: Emby.Naming/TV/SeriesPathParser.cs [hunk 11]
```diff
@@ -0,0 +1,95 @@
...
+                {
+                    if (int.TryParse(match.Groups[1].Value, NumberStyles.Integer, CultureInfo.InvariantCulture, out var num))
+                    {
+                        result.SeasonNumber = num;
+                    }
+                }
+
+                // Invalidate match when the season is 200 through 1927 or above 2500
+                // because it is an error unless the TV show is intentionally using false season numbers.
+                // It avoids erroneous parsing of something like "Series Special (1920x1080).mkv" as being season 1920 episode 1080.
```
Comment (cvium) [inline Emby.Naming/TV/SeriesPathParser.cs:84]: Should have a test for this

File: Emby.Naming/TV/SeriesResolver.cs [hunk 12]
```diff
@@ -0,0 +1,64 @@
...
+
+            if (!string.IsNullOrEmpty(seriesName))
+            {
+                seriesName = _seriesNameRegex.Replace(seriesName, "${a} ${b}").Trim();
+            }
+
+            return new SeriesInfo(path)
+            {
+                Name = seriesName,
+                SeasonNumber = seasonNumber
```
[the remark this answers is not in the document]
Comment (fredriklindberg) [reply Emby.Naming/TV/SeriesResolver.cs]: Fixed, removed the season number

File: Emby.Naming/TV/SeriesResolver.cs [hunk 13]
```diff
@@ -0,0 +1,64 @@
...
+    /// Used to resolve information about series from path.
+    /// </summary>
+    public class SeriesResolver
+    {
+        private readonly SeriesPathParser _seriesPathParser;
+
+        /// <summary>
+        /// Regex that matches strings of at least 2 characters separated by a dot or underscore.
+        /// </summary>
+        private static readonly Regex _seriesNameRegex = new Regex(@"((?<a>[^\._]{2,})[\._]*)|([\._](?<b>[^\._]{2,}))");
```
[the remark this answers is not in the document]
Comment (fredriklindberg) [reply Emby.Naming/TV/SeriesResolver.cs:17]: Turns things like "The_Show or "The.Show" into "The Show". But still preserves names like "S.H.O.W" or "Show A.B". Added a comment to clarify.

File: Emby.Naming/TV/SeriesResolver.cs (hunk 8 above)
Comment (fredriklindberg) [reply Emby.Naming/TV/SeriesResolver.cs]: Fixed

File: Emby.Naming/TV/SeriesPathParser.cs (hunk 9 above)
Comment (fredriklindberg) [reply Emby.Naming/TV/SeriesPathParser.cs]: Fixed

File: Emby.Naming/TV/SeriesInfo.cs (hunk 10 above)
Comment (fredriklindberg) [reply Emby.Naming/TV/SeriesInfo.cs]: Fixed, removed the season number

File: Emby.Naming/TV/SeriesPathParser.cs (hunk 11 above)
Comment (fredriklindberg) [reply Emby.Naming/TV/SeriesPathParser.cs]: This code block is gone since I removed the season numbers.

File: Emby.Naming/TV/SeriesPathParser.cs [hunk 14]
```diff
@@ -0,0 +1,64 @@
...
+            }
+
+            if (result != null)
+            {
+                if (!string.IsNullOrEmpty(result.SeriesName))
+                {
+                    result.SeriesName = result.SeriesName
+                        .Trim()
+                        .Trim('_', '.', '-')
+                        .Trim();
```
Comment (cvium) [inline Emby.Naming/TV/SeriesPathParser.cs:39]: Minor optimization thing (but given a few thousand items might add up), I wonder how often do we encounter special unicode whitespace chars? Would this `.Trim(' ', '_', '.', '-')` not cover something like 99% of cases?

Right now we're potentially creating 3 new strings for every file. The good thing about `.Trim()` is that it trims _all_ whitespace chars, but I can't help but wonder if it's even necessary.

Comment (fredriklindberg) [reply Emby.Naming/TV/SeriesPathParser.cs:39]: Probably good enough with  `.Trim(' ', '_', '.', '-')` yes

File: Emby.Naming/Common/NamingOptions.cs [hunk 15]
```diff
@@ -368,6 +368,20 @@ public NamingOptions()
                     IsOptimistic = true,
                     IsNamed = true
                 },
+
+                // Series and season only expression
+                // "the show/season 1", "the show/s01"
+                new EpisodeExpression(@"(.*(\\|\/))*(?<seriesname>.+)\/[Ss](eason){0,1}[\. _\-]*(?<seasonnumber>\d+)")
```
Comment (cvium) [inline Emby.Naming/Common/NamingOptions.cs:374]: We should always refrain from using `\d` as it covers a whole lot of unicode digits as well.
```suggestion
                new EpisodeExpression(@"(.*(\\|\/))*(?<seriesname>.+)\/[Ss](eason){0,1}[\. _\-]*(?<seasonnumber>[0-9]+)")
```

File: Emby.Naming/Common/NamingOptions.cs [hunk 16]
```diff
@@ -368,6 +368,20 @@ public NamingOptions()
...
+                {
+                    IsNamed = true
+                },
+
+                // Series and season only expression
+                // "the show S01", "the show season 1"
+                new EpisodeExpression(@"(.*(\\|\/))*(?<seriesname>.+)[\. _\-]+[sS](eason){0,1}[\. _\-]*(?<seasonnumber>\d+)")
```
Comment (cvium) [inline Emby.Naming/Common/NamingOptions.cs:381]:
```suggestion
                new EpisodeExpression(@"(.*(\\|\/))*(?<seriesname>.+)[\. _\-]+[sS](eason){0,1}[\. _\-]*(?<seasonnumber>[0-9]+)")
```

File: Emby.Naming/TV/SeriesPathParser.cs [hunk 17]
```diff
@@ -0,0 +1,64 @@
...
+            var result = new SeriesPathParserResult();
+
+            var match = expression.Regex.Match(name);
+
+            if (match.Success && match.Groups.Count >= 3)
+            {
+                if (expression.IsNamed)
+                {
+                    result.SeriesName = match.Groups["seriesname"].Value;
+                    result.Success = !string.IsNullOrEmpty(result.SeriesName) && !string.IsNullOrEmpty(match.Groups["seasonnumber"]?.Value);
```
Comment (cvium) [inline Emby.Naming/TV/SeriesPathParser.cs:57]: Does it matter if seasonnumber is empty?

Comment (fredriklindberg) [reply Emby.Naming/TV/SeriesPathParser.cs:54]: The path parser tries to be a bit "smart" in figuring out what's a series name or not, it does that by looking at common naming patterns. If the path parser fails the series resolver will just default back to the path name (as how it was before). 

So if there is a series path without any season identifier, the path will be returned by the resolver as is. See the tests for the series resolver.

Comment (fredriklindberg) [reply Emby.Naming/TV/SeriesPathParser.cs:54]: On second thought, the condition could be changed to be (episode number or season identifier). But would need some test for that to make it work.

File: Emby.Naming/TV/SeriesPathParser.cs (hunk 14 above)
Comment (fredriklindberg) [reply Emby.Naming/TV/SeriesPathParser.cs]: Fixed

File: Emby.Naming/Common/NamingOptions.cs (hunk 15 above)
Comment (fredriklindberg) [reply Emby.Naming/Common/NamingOptions.cs]: Fixed

File: Emby.Naming/Common/NamingOptions.cs (hunk 16 above)
Comment (fredriklindberg) [reply Emby.Naming/Common/NamingOptions.cs]: Fixed

**Review by cvium — APPROVED** [review]

**Review by crobibero — APPROVED** [review]

File: Emby.Naming/Common/NamingOptions.cs [hunk 18]
```diff
@@ -368,6 +368,20 @@ public NamingOptions()
...
+                // Series and season only expression
+                // "the show/season 1", "the show/s01"
+                new EpisodeExpression(@"(.*(\\|\/))*(?<seriesname>.+)\/[Ss](eason){0,1}[\. _\-]*(?<seasonnumber>[0-9]+)")
+                {
+                    IsNamed = true
+                },
+
+                // Series and season only expression
+                // "the show S01", "the show season 1"
+                new EpisodeExpression(@"(.*(\\|\/))*(?<seriesname>.+)[\. _\-]+[sS](eason){0,1}[\. _\-]*(?<seasonnumber>[0-9]+)")
```
Comment (Bond-009) [inline Emby.Naming/Common/NamingOptions.cs:381]:
```suggestion
                new EpisodeExpression(@"(.*(\\|\/))*(?<seriesname>.+)\/[Ss](eason)?[\. _\-]*(?<seasonnumber>[0-9]+)")
                {
                    IsNamed = true
                },

                // Series and season only expression
                // "the show S01", "the show season 1"
                new EpisodeExpression(@"(.*(\\|\/))*(?<seriesname>.+)[\. _\-]+[sS](eason)?[\. _\-]*(?<seasonnumber>[0-9]+)")
```

File: Emby.Naming/TV/SeriesPathParser.cs (hunk 17 above)
Comment (cvium) [reply Emby.Naming/TV/SeriesPathParser.cs:54]: Do you still want to do this?

Comment (fredriklindberg) [reply Emby.Naming/TV/SeriesPathParser.cs:54]: Do you have any sample data I can use for the test cases (I'm not familiar with anime)?

File: Emby.Naming/Common/NamingOptions.cs (hunk 18 above)
Comment (fredriklindberg) [reply Emby.Naming/Common/NamingOptions.cs]: Fixed
